src/number_theory/bernoulli_polynomials.lean

```diff
@@ -0,0 +1,238 @@
+/-
+Copyright (c) 2021 Ashvni Narayanan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Ashvni Narayanan
+-/
+
+import number_theory.bernoulli
+
+/-!
+# Bernoulli polynomials
+
+The Bernoulli polynomials are an important tool obtained from Bernoulli numbers.
+
+## Mathematical overview
+
+The $n^ℐrm{th}$ Bernoulli polynomial is defined as
+$$ B_n(X) = ∑_{k = 0}^n {n χoose k} (-1)^k * B_k * X^{n - k} $$
+where $B_k$ is the $k^ℐrm{th}$ Bernoulli number. The Bernoulli polynomials are generating functions,
+$$ t * e^{tX} / (e^t - 1) = ∑_{n = 0}^{∞} B_n(X) * ¼{t^n}{n!} $$
+
+## Implementation detail
+
+The Bernoulli (negative) numbers are first defined, `bernoulli_neg`, in order to make proving
+theorems using Bernoulli polynomials easier.
+
+## Main theorems
+
+`sum_bernoulli_neg : ∑ k in range n, (n.choose k : ℚ) * bernoulli_neg k = 0`
+`sum_bernoulli_poly : ∑ k in range (n + 1), ((n + 1).choose k : ℚ) • bernoulli_poly k = `
+  `polynomial.monomial n (n + 1 : ℚ)`
+`exp_bernoulli_poly : power_series.mk (λ n, (polynomial.aeval t ((1 / nat.factorial n : ℚ) • `
+`  bernoulli_poly n))) * (exp A - 1) = X * rescale t (exp A) `
+-/
+
+noncomputable theory
+open_locale big_operators
+open_locale nat
+
+namespace finset
+
+lemma sum_Ico_Ico_comm {M : Type*} [add_comm_monoid M]
+  (a b : ℕ) (f : ℕ → ℕ → M) :
+  ∑ i in finset.Ico a b, ∑ j in finset.Ico i b, f i j =
+  ∑ j in finset.Ico a b, ∑ i in finset.Ico a (j+1), f i j :=
+begin
+  rw ← @@finset.sum_sigma _ _ _ (λ i, finset.Ico i b) (λ x, f x.1 x.2),
+  rw ← @@finset.sum_sigma _ _ _ (λ j, finset.Ico a (j+1)) (λ x, f x.2 x.1),
+  refine finset.sum_bij'
+    (λ (x : Σ (i : ℕ), ℕ) _, (⟨x.2, x.1⟩ : Σ (i : ℕ), ℕ)) _ (λ _ _, rfl)
+    (λ (x : Σ (i : ℕ), ℕ) _, (⟨x.2, x.1⟩ : Σ (i : ℕ), ℕ)) _
+    (by rintro ⟨⟩ _; refl) (by rintro ⟨⟩ _; refl);
+  simp only [finset.Ico.mem, sigma.forall, finset.mem_sigma];
+  rintros a b ⟨⟨h₁,h₂⟩, ⟨h₃, h₄⟩⟩; refine ⟨⟨_, _⟩, ⟨_, _⟩⟩; linarith
+end
```

Comment:
The docs for `sum_sigma` say "use `sum_sigma'` when rewriting in the reverse direction:
```suggestion
lemma sum_Ico_Ico_comm {M : Type*} [add_comm_monoid M]
  (a b : ℕ) (f : ℕ → ℕ → M) :
  ∑ i in finset.Ico a b, ∑ j in finset.Ico i b, f i j =
  ∑ j in finset.Ico a b, ∑ i in finset.Ico a (j+1), f i j :=
begin
  rw finset.sum_sigma' _ (λ i, finset.Ico i b) f,
  rw finset.sum_sigma' _ (λ j, finset.Ico a (j+1)) (λ x y, f y x),
  refine finset.sum_bij'
    (λ (x : Σ (i : ℕ), ℕ) _, (⟨x.2, x.1⟩ : Σ (i : ℕ), ℕ)) _ (λ _ _, rfl)
    (λ (x : Σ (i : ℕ), ℕ) _, (⟨x.2, x.1⟩ : Σ (i : ℕ), ℕ)) _
    (by rintro ⟨⟩ _; refl) (by rintro ⟨⟩ _; refl);
  simp only [finset.Ico.mem, sigma.forall, finset.mem_sigma];
  rintros a b ⟨⟨h₁,h₂⟩, ⟨h₃, h₄⟩⟩; refine ⟨⟨_, _⟩, ⟨_, _⟩⟩; linarith
end
```

Reply:
In fact, even just `rw [finset.sum_sigma', finset.sum_sigma'],` works

Reply:
Done, thanks!